src/bin/pg_autoctl/pgctl.c

```diff
@@ -2046,6 +1917,69 @@ prepare_conninfo_sslmode(PQExpBuffer buffer, SSLOptions sslOptions)
 }
 
 
+/*
+ * pgctl_identify_system connects with replication=1 to our target node and run
+ * the IDENTIFY_SYSTEM command to check that HBA is ready.
+ */
+bool
+pgctl_identify_system(ReplicationSource *replicationSource)
+{
+	NodeAddress *primaryNode = &(replicationSource->primaryNode);
+
+	char primaryConnInfo[MAXCONNINFO] = { 0 };
+	char primaryConnInfoReplication[MAXCONNINFO] = { 0 };
+	PGSQL replicationClient = { 0 };
+	int len = 0;
+
+	/* sometimes we don't have a primary, just pretend everything is fine */
+	if (IS_EMPTY_STRING_BUFFER(primaryNode->host))
+	{
+		return true;
+	}
+
+	if (!prepare_primary_conninfo(primaryConnInfo,
+								  MAXCONNINFO,
+								  primaryNode->host,
+								  primaryNode->port,
+								  replicationSource->userName,
+								  NULL, /* no database */
+								  replicationSource->password,
+								  replicationSource->applicationName,
+								  replicationSource->sslOptions,
+								  false)) /* no need for escaping */
+	{
+		/* errors have already been logged. */
+		return false;
+	}
+
+	len = sformat(primaryConnInfoReplication, MAXCONNINFO,
+				  "%s replication=1",
+				  primaryConnInfo);
+
+	if (len >= MAXCONNINFO)
+	{
+		log_warn("Failed to call IDENTIFY_SYSTEM: primary_conninfo too large");
+		return false;
+	}
+
+	if (!pgsql_init(&replicationClient,
+					primaryConnInfoReplication,
+					PGSQL_CONN_UPSTREAM))
+	{
+		/* errors have already been logged */
+		return false;
+	}
+
+	if (!pgsql_identify_system(&replicationClient))
+	{
+		/* errors have already been logged */
+		return false;
+	}
+
```

Comment:
I guess it'd be  nice to use `pgsql_finish ` as we don't need the connection  anymore.

(We already fo  it in `pgsql_identify_system `, but still could be nicer to follow the code as we're injecting `replication=1`  and the readers could likely to expect  it  closed here)

Reply:
The connection is both opened and closed within `pgsql_identify_system` so it don't belongs to the upper layer in the pgctl module, the way I see it... which means I'm confused by your comment...